src/Workspaces/Core/Portable/CodeActions/CodeAction_Cleanup.cs

```diff
@@ -0,0 +1,188 @@
+        // good state. The semantic cleanup passes may have introduced new nodes with elastic trivia that have to be
+        // cleaned.
+        static (document, options, cancellationToken) => CleanupSyntaxAsync(document, options, cancellationToken),
+    ];
+
+    internal static ImmutableArray<DocumentId> GetAllChangedOrAddedDocumentIds(
+        Solution originalSolution,
+        Solution changedSolution)
+    {
+        var solutionChanges = changedSolution.GetChanges(originalSolution);
+        var documentIds = solutionChanges
+            .GetProjectChanges()
+            .SelectMany(p => p.GetChangedDocuments(onlyGetDocumentsWithTextChanges: true).Concat(p.GetAddedDocuments()))
+            .Concat(solutionChanges.GetAddedProjects().SelectMany(p => p.DocumentIds))
+            .ToImmutableArray();
+        return documentIds;
+    }
+
+    internal static async Task<Solution> CleanSyntaxAndSemanticsAsync(
+        Solution originalSolution,
+        Solution changedSolution,
+        CodeCleanupOptionsProvider optionsProvider,
+        IProgress<CodeAnalysisProgress> progress,
+        CancellationToken cancellationToken)
+    {
+        var documentIds = GetAllChangedOrAddedDocumentIds(originalSolution, changedSolution);
+        var documentIdsAndOptionsToClean = await GetDocumentIdsAndOptionsToCleanAsync().ConfigureAwait(false);
+
+        // Then do a pass where we cleanup semantics.
+        var cleanedSolution = await CleanupSyntaxAndSemanticsAsync(
+            changedSolution, documentIdsAndOptionsToClean, progress, cancellationToken).ConfigureAwait(false);
+
+        return cleanedSolution;
+
+        async Task<ImmutableArray<(DocumentId documentId, CodeCleanupOptions codeCleanupOptions)>> GetDocumentIdsAndOptionsToCleanAsync()
+        {
+            using var _ = ArrayBuilder<(DocumentId documentId, CodeCleanupOptions options)>.GetInstance(documentIds.Length, out var documentIdsAndOptions);
+            foreach (var documentId in documentIds)
+            {
+                var document = changedSolution.GetRequiredDocument(documentId);
+
+                // Only care about documents that support syntax.  Non-C#/VB files can't be cleaned.
+                if (document.SupportsSyntaxTree)
+                {
+                    var codeActionOptions = await document.GetCodeCleanupOptionsAsync(optionsProvider, cancellationToken).ConfigureAwait(false);
+                    documentIdsAndOptions.Add((documentId, codeActionOptions));
+                }
+            }
+
+            return documentIdsAndOptions.ToImmutableAndClear();
+        }
+    }
+
+    internal static async ValueTask<Document> CleanupDocumentAsync(Document document, CodeCleanupOptions options, CancellationToken cancellationToken)
+    {
+        if (!document.SupportsSyntaxTree)
+            return document;
+
+        var cleanedSolution = await CleanupSyntaxAndSemanticsAsync(
```

Comment:
cleaning a document just defers to the main N-pass helper, just processing this single document.